init.lua

```diff
@@ -1,16 +1,81 @@
+local F = minetest.formspec_escape
+local C = minetest.colorize
 
 minetest.register_alias("castle:workbench", "crafting_bench:workbench")
 
+local has_mcl = minetest.get_modpath("mcl_formspec")
+local has_default = minetest.get_modpath("default")
+
 local usage_help = S("The inventory on the left is for raw materials, the inventory on the right holds finished products. The crafting grid in the center defines what recipe this workbench will make use of; place raw materials into it in the crafting pattern corresponding to what you want to build.")
 
-if minetest.get_modpath("hopper") and hopper ~= nil and hopper.add_container ~= nil then
+if ( minetest.get_modpath("hopper") and hopper ~= nil and hopper.add_container ~= nil ) or has_mcl then
 	usage_help = usage_help .. "\n\n" .. S("This workbench is compatible with hoppers. Hoppers will insert into the raw material inventory and remove items from the finished goods inventory.")
 end
 
 
 local crafting_rate = tonumber(minetest.settings:get("crafting_bench_crafting_rate")) or 5
 
+if not has_default and not has_mcl then
+	error("The crafting bench mod needs either the default mod (minetest game) or mineclonia/mineclone2 to work")
+end
+
+local invsize_src = 2 * 4
+local invsize_dst = 1 * 4
+
+if has_mcl then
+	invsize_src = 2 * 3
+	invsize_dst = 1 * 3
+end
+
+local groups
+local mcl_hardness
+local mcl_blast_res
+
+local formspec
+if has_default then
+	formspec = 'size[10,10;]' ..
+		default.gui_bg ..
+		default.gui_bg_img ..
+		default.gui_slots ..
+		'label[1,0;'..S('Source Material')..']' ..
+		'list[context;src;1,1;2,4;]' ..
+		'label[4,0;'..S('Recipe to Use')..']' ..
+		'list[context;rec;4,1;3,3;]' ..
+		'label[7.5,0;'..S('Craft Output')..']' ..
+		'list[context;dst;8,1;1,4;]' ..
+		'list[current_player;main;1,6;8,4;]' ..
+		'listring[current_name;dst]'..
+		'listring[current_player;main]'..
+		'listring[current_name;src]'..
+		'listring[current_player;main]'
+	groups = {choppy=2, oddly_breakable_by_hand=2, flammable=2}
+elseif has_mcl then
+	formspec ="formspec_version[4]"..
+		"size[11.75,10.425]"..
+
+		mcl_formspec.get_itemslot_bg_v4(1, 0.75, 2, 3)..
+		mcl_formspec.get_itemslot_bg_v4(5, 0.75, 3, 3)..
+		mcl_formspec.get_itemslot_bg_v4(10, 0.75, 1, 3)..
+		'label[1,0.375;'..S('Source Material')..']' ..
+		'list[context;src;1,0.75;2,4;]' ..
+		'label[5,0.375;'..S('Recipe to Use')..']' ..
+		'list[context;rec;5,0.75;3,3;]' ..
+		'label[9.5,0.375;'..S('Craft Output')..']' ..
+		'list[context;dst;10,0.75;1,4;]' ..
+
+		"label[0.375,4.7;" .. F(C(mcl_formspec.label_color, S("Inventory"))) .. "]"..
+		mcl_formspec.get_itemslot_bg_v4(0.375, 5.1, 9, 3)..
+		"list[current_player;main;0.375,5.1;9,3;9]"..
+		mcl_formspec.get_itemslot_bg_v4(0.375, 9.05, 9, 1)..
+		"list[current_player;main;0.375,9.05;9,1;]"..
+
+		"listring[current_player;src]"..
+		"listring[current_player;main]"
+	groups = {axey=2, handy=2, flammable=-1, container = 4}
+	mcl_hardness = 2
+	mcl_blast_res = 3
+end
```

Comment:
What if neither is present? A specific `error` call would be very welcome to not end up with arbitrary "access to nil value" errors or whatsoever.